contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,213 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {IWarpMessenger} from "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import {INativeMinter} from "@subnet-evm-contracts/interfaces/INativeMinter.sol";
+import {INativeTokenDestination} from "./INativeTokenDestination.sol";
+import {ITeleporterMessenger, TeleporterFeeInfo, TeleporterMessageInput} from "../../Teleporter/ITeleporterMessenger.sol";
+import {ITeleporterReceiver} from "../../Teleporter/ITeleporterReceiver.sol";
+import {SafeERC20TransferFrom} from "../../Teleporter/SafeERC20TransferFrom.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+// We need IAllowList is an indirect dependency in order to compile.
+// solhint-disable-next-line
+import {IAllowList} from "@subnet-evm-contracts/interfaces/IAllowList.sol";
+
+// The address where the burned transaction fees are credited.
+// TODO implement mechanism to report burned tx fees to source chian.
+address constant BURNED_TX_FEES_ADDRESS = 0x0100000000000000000000000000000000000000;
+// Designated Blackhole Address. Tokens are sent here to be "burned" before sending an unlock
+// message to the source chain. Different from the burned tx fee address so they can be
+// tracked separately.
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000001;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(0x0200000000000000000000000000000000000001);
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 100_000;
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable nativeTokenSourceAddress;
+    // The first `initialReserveImbalance` tokens sent to this subnet will not be minted.
+    // `initialReserveImbalance` should be constructed to match the initial token supply of this subnet.
+    // This means tokens will not be minted until the source contact is collateralized.
+    uint256 public immutable initialReserveImbalance;
+    uint256 public currentReserveImbalance;
+    uint256 public totalMinted = 0;
+
+    // Used for sending and receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
```

Comment:
Now that the registry and upgrade strategy are more complete, we should consider if we want to integrate here. The launch plan and initial bootstrapping would have changes if yes.

Reply:
I've implemented the changes, but I'm actually going to hold off on integrating them until the e2e tests have been set up to deploy the registry